### ISSUE TITLE
Feature Request: Add support for recursive checking of inner exception

### ISSUE THREAD

**Issue by momt99** [fluentassertions/fluentassertions#1751]
### Description
Sometimes, when writing high-level tests, we expect a function to throw an exception with a deeply nested inner exception. Writing assertions can be tedious if we only care about the existence of a certain type in the chain and not how deep it is located.

So in short, I would like to write something like this:
```csharp
action.Throws<ExceptionA>()
    .WithInnerException<ExceptionB>(recursive: true);
```
instead of:
```csharp
action.Throws<ExceptionA>()
    .Which.InnerException.InnerException.InnerException.BeOfType<ExceptionB>();
```

**Comment by momt99** [comment]
Actually, I have written an extension method for this functionality. If this is correct, I can submit a pull request.
```csharp
public static ExceptionAssertions<TInnerException> WithInnerException<TException, TInnerException>(
    this ExceptionAssertions<TException> exceptionAssertions,
    bool recursive,
    string because = "",
    params object[] becauseArgs)
    where TException : Exception
    where TInnerException : Exception
{
    Execute.Assertion
        .BecauseOf(because, becauseArgs)
        .WithExpectation("Expected inner {0}{reason}, but ", typeof(TInnerException))
        .ForCondition(exceptionAssertions.Subject is not null)
        .FailWith("no exception was thrown.")
        .Then
        .ForCondition(exceptionAssertions.Subject!.Any(e => e.InnerException is not null))
        .FailWith("the thrown exception has no inner exception.")
        .Then
        .ClearExpectation();

    TInnerException[] expectedInnerExceptions =
        (recursive
            ? exceptionAssertions.Subject!
                .Select(e => GetRecursiveInnerExceptions(e).OfType<TInnerException>().FirstOrDefault())
            : exceptionAssertions.Subject!
                .Select(e => e.InnerException))
        .OfType<TInnerException>()
        .ToArray();

    Execute.Assertion
        .ForCondition(expectedInnerExceptions.Any())
        .BecauseOf(because, becauseArgs)
        .FailWith(
            "Expected inner {0}{reason}, but found {1}.",
            typeof(TInnerException),
            exceptionAssertions.Subject!.Single().InnerException);

    return new ExceptionAssertions<TInnerException>(expectedInnerExceptions);

    static IEnumerable<Exception> GetRecursiveInnerExceptions(
        Exception ex)
    {
        var current = ex.InnerException;
        while (current is not null)
        {
            yield return current;
            current = current.InnerException;
        }
    }
}
```

**Comment by dennisdoomen** [comment]
I'm trying to come up with a good name for that overload. Something like `WithInnerExceptionAnywhere`. Ideally we would want to make that the default, but then you would need a `WithDirectInnerException`. And that would be a breaking change as well.

**Comment by jnyrup** [comment]
Interesting use case, I definitely see why you want a higher-level assertion.
I'm not sure the recursive search should be the default behavior of `WithInnerException`, as it might be slightly misleading for those who want to assertion on the _immediate_ inner exception.

**Comment by eNeRGy164** [comment]
Maybe look at XLinq type of naming? Something like `WithDescendantException`

**Comment by tullo-x86** [comment]
Would `action.Should().Throw<ExceptionA>().ContainingInnerException<ExceptionB>()` make sense?

**Comment by dennisdoomen** [comment]
More ideas through [this Twitter feed](https://twitter.com/ddoomen/status/1465302319420121090):
* `WithNestedInnerException`
* `WithDeepInnerException`

**Comment by markusschaber** [comment]
Maybe the code should also consider `AggregateException`s which have an additional collection member `InnerExceptions`.

**Comment by julealgon** [comment]
Another possibility for this would be to convert it into a normal collection assertion with something like:
```csharp
action.Should().Throw<ExceptionA>().WhoseInnerExceptionChain.Should().Contain<ExceptionB>();
```

This would be similar to the `WhoseValue` used in places like dictionaries and leverage all the collection assertions that already exist on top of the flattened list of exceptions (that could also be empty). It could also be a computed `IEnumerable` (i.e. deferred execution), depending on how we want to go about it.

I think the "Chain" wording at the end makes sense, but perhaps there is a better way to put it (perhaps just `WhoseInnerExceptions`, even though that could be a bit misleading when dealing with an `AggregateException`).

From what I'm seeing in the docs, it seems we don't have a generic `Contain<T>` assertion, so I'd probably add that as well to make this work, since it would be a valid assertion to have for normal collections too (would be useful for polymorphic collections). Naming could either be `Contain<T>`, or maybe a more explicit `ContainItemOfType<T>` to go along the other "Contain...Items..." methods. While we are at it, a "Single" version should also be added to match.

**Comment by dennisdoomen** [comment]
I think that's a great idea and I would prefer `WhoseInnerExceptionChain` to emphasize that that it exposes not just the direct inner exceptions.

**Comment by ITaluone** [comment]
Hi

Is this ticket api-approved and can be implemented? Or do we need something else here?

**Comment by ITaluone** [comment]
My proposed API changes are:

```diff
GenericCollectionAssertions<T>
{
+         public FluentAssertions.AndConstraint<TAssertions> ContainItemsOfType<TExpectation>(string because = "", params object[] becauseArgs) { }
}
ExceptionAssertions<TException>
{
+         public System.Collections.Generic.IEnumerable<TException> WhoseInnerExceptionChain { get; }
}
```

But I'm kind of unsure, wether name it `ContainItemsOfType` or `ContainItemsOfExactType`..

**Comment by dennisdoomen** [comment]
I'm all for `WhoseInnerExceptionChain`, but I don't think we should add those two `ContainXX` methods considering how [many of those assertion methods](https://github.com/fluentassertions/fluentassertions/blob/62f0d768a802fbfe630c887afe593b7680eeffcb/Tests/Approval.Tests/ApprovedApi/FluentAssertions/netcoreapp3.0.verified.txt#L414) we already have, including `ContainItemsAssignableTo`.

**Comment by IT-VBFK** [comment]
hmm..

What about giving `ContainItemsAssignableTo` an overload which check the type exactly?

As I understand the `ContainItemsAssignableTo`: it matches also derivatives of a type `T`, but I don't think this is what we want here, do we?

**Comment by dennisdoomen** [comment]
Yeah, that's true. But then you need `ContainItemsOfExactType` and `ContainSingleOfExactType`.

**Comment by ITaluone** [comment]
That's why I proposed ths two new `Contain...<T>`  methods.

What about something like:
```csharp

action.Should().Throw<ExceptionA>().WhoseInnerExceptionChain.Should().HaveElementsOfExactType<ExceptionB>();

// or

action.Should().Throw<ExceptionA>().WhoseInnerExceptionChain.Should().HaveElementsOfType<ExceptionB>();

// or

action.Should().Throw<ExceptionA>().WhoseInnerExceptionChain.Should().HaveItemsOfType<ExceptionB>();

```

Using `GenericCollectionAssertions` here makes the most sense IMHO.

**Comment by IT-VBFK** [comment]
Another approach could be:

```csharp

public class ExceptionChain
{
        internal IEnumerable<Exception> elements;

        public Has<T>();
}

public class ExceptionAssertion<TException>
{
        public ExceptionChain WhoseInnerExceptionChain { get; }
}

action.Should().Throw<ExceptionA>().WhoseInnerExceptionChain.Has<ExceptionB>();

```

**Comment by dennisdoomen** [comment]
I prefer `HaveElementsOfType<ExceptionB>`

**Comment by IT-VBFK** [comment]
Ok

And for the single overload? (If we address this here too)

```csharp

action.Should().Throw<ExceptionA>().WhoseInnerExceptionChain.Should().HaveElementOfType<ExceptionB>();

```

**Comment by julealgon** [comment]
> Or using the neat little thing called `OccurrenceConstraint` 😎
> 
> so this would then be for example:
> 
> ```cs
> action.Should().Throw<ExceptionA>().WhoseInnerExceptionChain.Should().HaveElementsOfType<ExceptionB>(Exactly.Once()); // for both
> ```
> 
> But in this case I would use `HaveElementOfType<Foo>(Exactly.Once());`... reads more fluent IMHO

Is this actually idiomatic FluentAssertions? I don't remember seeing it in other overloads but I might just be misremembering.

**Comment by IT-VBFK** [comment]
Was introduced in #1145 3 years ago

**Comment by julealgon** [comment]
> Was introduced in #1145 3 years ago

That's..... very surprising. Makes me wonder why it is used so sparingly. I imagine some methods could be deprecated in favor of that even.

**Comment by dennisdoomen** [comment]
> Or using the neat little thing called `OccurrenceConstraint`

Hmm, that might not be a bad idea indeed. Wondering what @jnyrup thinks about that.

**Comment by jnyrup** [comment]
If we could change `ContainItemsAssignableTo` to return an `AndWhichConstraint<TAssertions, IEnumerable<TExpectation>>` we could do this.

```cs
action.Should().Throw<Exception>()
    .WhoseInnerExceptionChain.Should().ContainItemsAssignableTo<InvalidOperationException>()
        .Which.Should().HaveCount(42);
```
But I'm not sure this is possible since we also support checking whether a `Type` is assignable to another `Type`.
And it doesn't read _that_ well.

Another possibility would be to add an overload of `ContainItemsAssignableTo` that took an `OccurrenceConstraint`.
```cs
action.Should().Throw<Exception>()
    .WhoseInnerExceptionChain.Should().ContainItemsAssignableTo<InvalidOperationException>(Exactly.Times(42));
```

Some more brainstorming on a name for `WhoseInnerExceptionChain` as I'm not (yet) completely sold on the "Chain" part of the name.
* `WhoseNestedExceptions` 
* `WhoseNestedInnerExceptions`
* `WhoseFlattenedExceptions` (inspired by `AggregateException.Flatten()`)

**Comment by dennisdoomen** [comment]
I like `WhoseInnerExceptionChain` and `WhoseNestedException` equally, but I don't like the use of `ContainItemsAssignableTo`. Not just because it doesn't read very nicely, but also because we most likely care about the exact type.

**Comment by IT-VBFK** [comment]
Hmm.. that limits the possibilities of a fluent syntax extremely..

But what if we omit the `Should()` after `WhoseInnerExceptionChain`?

So for example:

`action.Should().Throw<Exception>()
    .WhoseInnerExceptionChain.ContainsWithExactType<InvalidOperationException>(Exactly.Times(42));`

or something like this..

**Comment by ITaluone** [comment]
Can you @dennisdoomen @jnyrup please clarify wether the way of finally asserting a generic collection is the right one.

or should we concentrate on asserting only a `IEnumerable<Exception>` here?

**Comment by jnyrup** [comment]
@ITaluone I'm not sure what you're asking for?
I'm still in a brain storming phase exploring different shapes for the feature, both in terms of functionality and naming.

**Comment by IT-VBFK** [comment]
What I meant was: Should we concentrate only on exceptions (and inner exceptions) or should this feature also affect all collections.

**Comment by jnyrup** [comment]
If `WhoseInnerExceptionChain` returns an `IEnumerable<Exception>` and we want to be able to write
`action.Should().Throw<SomeException>().WhoseInnerExceptionChain.Should().XYZ()` I guess we have to look at both exceptions and collections in general at the same time.

Maybe except if we make `XYZ()` a dedicated extension method on top of `GenericCollectionAssertions<T> where T : Exception`.

**Comment by IT-VBFK** [comment]
I think a dedicated extension method would fit better into the structure of the `GenericCollectionAsstertions` to not pollute this class, because now I would have to take care if the type is an exception.

So I have to do this:

```

if(T is Exception) 
{
        Assert.Exceptions...
}
else // any other type
{
        Assert.AnyOtherType
}

```

And writing `where T : Exception` inside the `GenericColltionAssertions` is not quite generic :D

**Comment by ITaluone** [comment]
See comment https://github.com/fluentassertions/fluentassertions/issues/1716#issuecomment-950374781

In my opinion this is a niche :)

**Comment by skwasjer** [comment]
I have not seen consideration of the async (eg. `Func<Task>`) exception assertion API yet which does not work well with Whose/Which property API and requires use of extension regardless. To keep things consistent I would avoid using a property to chain.

Considering it is somewhat of a niche use case I vote for proposal by OP, or otherwise use an (optional) options delegate pattern.
```csharp
Action action = () => {};
Func<Task> func = () => Task.Completed;
action.Should().Throw<SomeException>().WithInnerException<InnerException>(opts => opts.Recursive = true);
await func.Should().ThrowAsync<SomeException>().WithInnerException<InnerException>(opts => opts.Recursive = true);
```

**Comment by julealgon** [comment]
> I have not seen consideration of the async (eg. `Func<Task>`) exception assertion API yet which does not work well with Whose/Which property API and requires use of extension regardless.

I don't think this is necessarily the case. One could create a custom awaitable class that wraps the original task, awaits on it, while still providing the properties.

It is not a very common thing, sure, but it should definitely be possible.
